lib/plug/cowboy.ex

```diff
@@ -125,6 +125,33 @@ defmodule Plug.Cowboy do
 
   To opt-out of this default instrumentation, you can manually configure
   cowboy with the option `stream_handlers: [:cowboy_stream_h]`.
+
+  ## WebSocket support
+
+  Plug.Cowboy supports upgrading HTTP requests to WebSocket connections via 
+  the use of the `Plug.Conn.upgrade_adapter/3` function, called with `:websocket` as the second
+  argument. Applications should validate that the connection represents a valid WebSocket request
+  before calling this function (Cowboy will validate the connection as part of the upgrade
+  process, but does not provide any capacity for an application to be notified if the upgrade is
+  not successful). If an application wishes to negotiate WebSocket subprotocols or otherwise set
+  any response headers, it should do so before calling `Plug.Conn.upgrade_adapter/3`.
+
+  The third argument to `Plug.Conn.upgrade_adapter/3` defines the details of how Plug.Cowboy
+  should handle the WebSocket connection, and must take the form `{handler, handler_opts,
+  connection_opts}`, where values are as follows:
```

Comment:
I am thinking the `connection_opts` should just be the Cowboy options and we should not rewrite it in any way. So we just pass it as a map to Cowboy, as it expects, and we do not touch it. WDYT?

Reply:
`connection_opts` is as-defined by the options in `Phoenix.Socket`, so it would need to change there (I didn't pick it randomly)

Reply:
Yes, I am aware, Phoenix would need to translate its own options to Cowboy options at some point, but we can do it! We are just moving it up a bit!

Reply:
Done